pkg/base/k8s/cr_helper.go

```diff
@@ -105,29 +106,61 @@ func (cs *CRHelper) DeleteACsByNodeID(nodeID string) error {
 	return nil
 }
 
-// GetVolumeByLocation reads the whole list of Volume CRs from a cluster and searches the volume with provided location
+// GetVolumesByLocation reads the whole list of Volume CRs from a cluster and searches the volume with provided location
 // Receives golang context and location name which should be equal to Volume.Spec.Location
-// Returns a pointer to the instance of volumecrd.Volume or nil
-func (cs *CRHelper) GetVolumeByLocation(location string) *volumecrd.Volume {
+// Returns a list of a pointers to volumes which are belong to the location and error
+func (cs *CRHelper) GetVolumesByLocation(ctx context.Context, location string) ([]*volumecrd.Volume, error) {
 	ll := cs.log.WithFields(logrus.Fields{
-		"method":   "GetVolumeByLocation",
+		"method":   "GetVolumesByLocation",
 		"location": location,
 	})
 
+	var volumes []*volumecrd.Volume
 	volList := &volumecrd.VolumeList{}
-	if err := cs.k8sClient.ReadList(context.Background(), volList); err != nil {
+	if err := cs.k8sClient.ReadList(ctx, volList); err != nil {
 		ll.Errorf("Failed to get volume CR list, error %v", err)
-		return nil
+		return nil, err
 	}
-
+	var lvgUUID *string
 	for _, v := range volList.Items {
+		v := v
+		if v.Spec.LocationType == apiV1.LocationTypeLVM {
+			// lazy read
+			if lvgUUID == nil {
+				lvgUUID = cs.getLVGByDrive(ctx, location)
```

Comment:
I can't understand that logic, why do you use `location` that was came as a parameter, as a location for each volume from list?

Reply:
Location is a drive ID. In this function we are always looking for a volumes which are located on that drive. If the volume is a "full-disk", we need to compare v.Spec.Location with location from function arg (drive ID). If volume is "LVM" based we can't use v.Spec.Location to filter volumes, because for such volumes LVG id stored in v.Spec.Location instead of drive id.  So, we need to find LVG which located on required drive and then filter volumes by the LVG id.
If we have no LVG based volumes we don't need to read LVG list at all, so we use lazy read here. We will read LVG list only once, when we find first lvm based volume (resolved lvgUUID will be placed to location var)

Reply:
Still not clear for me, lets say we have next scenario:
- there are LVG-based on HDD-based volumes in the cluster, e.g. `vol1` - it is LVM - based volume, `vol2` - HDD
- function GetVolumesByLocation was called with `location` that points on HDD-based volume and we expect to get `vol2`, according for our example
- function read list of all volumes: `[vol1, vol2]`
- first loop iteration use `vol1`, as it was said before `vol1` is LVM-based 
- function run `cs.getLVGByDrive(ctx, location)` that return nil for current example
- and in line 132: `if lvgUUID == nil` function gets `True` and returns error

Reply:
You are right. Nice finding! 👍 Thank you! Fixed